src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -178,9 +191,21 @@ private void WriteHeaders(HttpRequestMessage request)
                 WriteHeader(ref state, ":authority", authority);
                 WriteHeader(ref state, ":path", request.RequestUri.GetComponents(UriComponents.PathAndQuery | UriComponents.Fragment, UriFormat.UriEscaped));
 
-                if (request.HasHeaders)
+                // Determine cookies to send.
+                string cookiesFromContainer = null;
+                if (_connection.UseCookie)
                 {
-                    WriteHeaders(ref state, request.Headers);
+                    cookiesFromContainer = _connection.GetCookieContainer.GetCookieHeader(request.RequestUri);
+                    if (cookiesFromContainer == "")
+                    {
+                        cookiesFromContainer = null;
+                    }
+                }
+
+                // Write request headers.
+                if (request.HasHeaders || cookiesFromContainer != null)
+                {
+                    WriteHeaders(ref state, request.Headers, cookiesFromContainer);
```

Comment:
What about manual 'Cookie' headers added to the request by the developer?  How are those being processed and sent?  Are we following the current SocketsHttpHandler behavior on HTTP/1.1 w.r.t. manually-added 'Cookie' headers? I don't recall the exact behaviors especially if both CookieContainer/UseCookies is on and manually-added headers are set.

Reply:
> What about manual 'Cookie' headers added to the request by the developer?

They will be wrote to request header and sent.

> How are those being processed and sent?

Yes. We write Cookies from CookieContainer first. Then we add manually-added cookies.
[HTTP/1.1 Code.](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L236-L260)
[HTTP/2 Code - newly added.](https://github.com/caesar1995/corefx/blob/09138d5aa6f045faac200016347821f17f639b24/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L135-L146)


> Are we following the current SocketsHttpHandler behavior on HTTP/1.1 w.r.t. manually-added 'Cookie' headers?

Yes.

> I don't recall the exact behaviors especially if both CookieContainer/UseCookies is on and manually-added headers are set.

Both from CookieContainer and manually-added headers will be sent. This is true for both HTTP/1.1 and HTTP/2.
HTTP/1.1 reference: https://github.com/dotnet/corefx/issues/26983#issuecomment-384755766
For HTTP/2, here is the test code (for reference): https://github.com/dotnet/corefx/pull/31440/files#diff-f54b77e678f81d428a2708760a67ba93R195

Reply:
Here is the test to verify that both Cookies from CookieContainer and CookieHeader will be sent on HTTP/1.1 (for SocketsHttpHandler & WinHttpHandler):

https://github.com/dotnet/corefx/blob/4e91b8620498cc82faafecb619bca630e277d6d3/src/System.Net.Http/tests/FunctionalTests/HttpCookieProtocolTests.cs#L195